src/python/pants/engine/internals/graph.py

```diff
@@ -98,85 +98,88 @@ async def resolve_unexpanded_targets(addresses: Addresses) -> UnexpandedTargets:
 
 
 @rule
-async def generate_subtargets(address: Address) -> Subtargets:
-    if address.is_file_target:
-        raise ValueError(f"Cannot generate file Targets for a file Address: {address}")
-
-    wrapped_build_target = await Get(WrappedTarget, Address, address)
-    build_target = wrapped_build_target.target
-
-    supports_subtargets = True
-    if build_target.has_field(Sources):
-        sources_field = build_target.get(Sources)
-        supports_subtargets = not sources_field.indivisible
-    else:
-        supports_subtargets = False
-
-    if not build_target.has_field(Dependencies):
-        supports_subtargets = False
-
-    if not supports_subtargets:
-        # If a target type does not support dependencies or has no Sources field (or the Sources field is
-        # indivisible), we do not split it, as that would prevent the BUILD target from depending on its splits.
-        return Subtargets(build_target, ())
-
-    # Generate a subtarget per source.
-    paths = await Get(SourcesPaths, SourcesPathsRequest(build_target[Sources]))
-    wrapped_subtargets = await MultiGet(
-        Get(
-            WrappedTarget,
-            Address,
-            generate_subtarget_address(address, full_file_name=subtarget_file),
-        )
-        for subtarget_file in paths.files
+def target_types_to_generate_targets_requests(
+    union_membership: UnionMembership,
+) -> TargetTypesToGenerateTargetsRequests:
+    return TargetTypesToGenerateTargetsRequests(
+        {
+            request_cls.generate_from: request_cls  # type: ignore[misc]
+            for request_cls in union_membership.get(GenerateTargetsRequest)
+        }
     )
-    return Subtargets(build_target, tuple(wt.target for wt in wrapped_subtargets))
 
 
 @rule
 async def resolve_target(
     address: Address,
     registered_target_types: RegisteredTargetTypes,
     union_membership: UnionMembership,
+    target_types_to_generate_requests: TargetTypesToGenerateTargetsRequests,
 ) -> WrappedTarget:
     if address.is_file_target:
```

Comment:
There is enough inside this block now that it seems like the cases should be inverted: a BUILD target is now the fast / early-return path.